File: src/Blazor/Movies/Services/MoviesService.cs
using System.Text.Json;
using Blazor.Core.Models;
using Blazor.Movies.Requests;
using Blazor.Shared.Results;

namespace Blazor.Movies.Services;

public sealed class MoviesService : IMoviesService
{
    private readonly HttpClient _httpClient;
    private const string MoviesRoute = "movies/";

    private static readonly JsonSerializerOptions SerializerOptions = new()
    {
        PropertyNameCaseInsensitive = true
    };

    public MoviesService(HttpClient httpClient)
    {
        _httpClient = httpClient;
    }

    public async Task<OneOf<List<Movie>, Error>> GetMovies()
    {
        var response = await _httpClient.GetAsync(MoviesRoute);
        if (!response.IsSuccessStatusCode)
        {
            var error = await response.Content.ReadFromJsonAsync<Error>(SerializerOptions);
            return error ?? new Error("Failed to load movies.");
        }

        var movies = await response.Content.ReadFromJsonAsync<List<Movie>?>(SerializerOptions);

        return movies ?? [];
    }

    public async Task<OneOf<Success, Error>> CreateMovie(CreateMovieRequest request)
    {
        var httpContent = JsonContent.Create(request, options: SerializerOptions);
        var response = await _httpClient.PostAsync(MoviesRoute, httpContent);

        if (response.IsSuccessStatusCode) return StaticResults.Success;

        var error = await response.Content.ReadFromJsonAsync<Error>(SerializerOptions);
        return error ?? new Error("Failed to create movie.");
    }

    public async Task<OneOf<Success, Error>> UpdateMovie(Guid movieId, UpdateMovieRequest request)
    {
        var httpContent = JsonContent.Create(request, options: SerializerOptions);
        var response = await _httpClient.PutAsync(MoviesRoute + movieId, httpContent);

        if (response.IsSuccessStatusCode) return StaticResults.Success;

        var error = await response.Content.ReadFromJsonAsync<Error>(SerializerOptions);
        return error ?? new Error($"Failed to update movie.");
    }

    public async Task<OneOf<Success, Error>> DeleteMovie(Guid movieId)
    {
        var response = await _httpClient.DeleteAsync(MoviesRoute + movieId);

        if (response.IsSuccessStatusCode) return StaticResults.Success;

        var error = await response.Content.ReadFromJsonAsync<Error>(SerializerOptions);
        return error ?? new Error($"Failed to delete movie.");
    }

    public async Task<OneOf<Success, Error>> ReturnMovie(Guid movieId, Guid clientId)
    {
        var response = await _httpClient.PostAsync($"{MoviesRoute}{movieId}/return?clientId={clientId}", null);

        if (response.IsSuccessStatusCode) return StaticResults.Success;

        var error = await response.Content.ReadFromJsonAsync<Error>(SerializerOptions);
        return error ?? new Error("Failed to return movie.");
    }
}

File: src/Svelte/src/lib/services/movies.service.ts
import { PUBLIC_API_URL } from '$env/static/public';
import type { Movie } from '$lib/types/movie.model';
import type { Result } from '$lib/types/result.model';
export interface CreateMovieRequest {
    title: string;
    description: string;
    pricePerDay: number;
    collateralValue: number;
}

export interface UpdateMovieRequest {
    title: string;
    description: string;
    pricePerDay: number;
    collateralValue: number;
}

class MoviesService {
    private baseUrl = `${PUBLIC_API_URL}/movies`;

    async getMovies(): Promise<Result<Movie[]>> {
        try {
            const response = await fetch(this.baseUrl);

            if (!response.ok) {
                const error = await response.json().catch(() => ({ message: 'Failed to load movies.' }));
                return { success: false, error };
            }

            const movies = await response.json();
            return { success: true, data: movies ?? [] };
        } catch (err) {
            return { success: false, error: { message: 'Failed to load movies.' } };
        }
    }

    async getMovieById(movieId: string): Promise<Result<Movie>> {
        try {
            const response = await fetch(`${this.baseUrl}/${movieId}`);

            if (!response.ok) {
                const error = await response.json().catch(() => ({ message: 'Failed to load movie details.' }));
                return { success: false, error };
            }

            const movie = await response.json();

            if (!movie) {
                return { success: false, error: { message: 'Failed to deserialize movie.' } };
            }

            return { success: true, data: movie };
        } catch (err) {
            return { success: false, error: { message: 'Failed to load movie details.' } };
        }
    }

    async createMovie(request: CreateMovieRequest): Promise<Result<void>> {
        try {
            const response = await fetch(this.baseUrl, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                },
                body: JSON.stringify(request),
            });

            if (!response.ok) {
                const error = await response.json().catch(() => ({ message: 'Failed to create movie.' }));
                return { success: false, error };
            }

            return { success: true, data: undefined };
        } catch (err) {
            return { success: false, error: { message: 'Failed to create movie.' } };
        }
    }

    async updateMovie(movieId: string, request: UpdateMovieRequest): Promise<Result<void>> {
        try {
            const response = await fetch(`${this.baseUrl}/${movieId}`, {
                method: 'PUT',
                headers: {
                    'Content-Type': 'application/json',
                },
                body: JSON.stringify(request),
            });

            if (!response.ok) {
                const error = await response.json().catch(() => ({ message: 'Failed to update movie.' }));
                return { success: false, error };
            }

            return { success: true, data: undefined };
        } catch (err) {
            return { success: false, error: { message: 'Failed to update movie.' } };
        }
    }

    async deleteMovie(movieId: string): Promise<Result<void>> {
        try {
            const response = await fetch(`${this.baseUrl}/${movieId}`, {
                method: 'DELETE',
            });

            if (!response.ok) {
                const error = await response.json().catch(() => ({ message: 'Failed to delete movie.' }));
                return { success: false, error };
            }

            return { success: true, data: undefined };
        } catch (err) {
            return { success: false, error: { message: 'Failed to delete movie.' } };
        }
    }

    async returnMovie(movieId: string, clientId: string): Promise<Result<void>> {
        try {
            const response = await fetch(`${this.baseUrl}/${movieId}/return?clientId=${clientId}`, {
                method: 'POST',
            });

            if (!response.ok) {
                const error = await response.json().catch(() => ({ message: 'Failed to return movie.' }));
                return { success: false, error };
            }

            return { success: true, data: undefined };
        } catch (err) {
            return { success: false, error: { message: 'Failed to return movie.' } };
        }
    }
}

export const moviesService = new MoviesService();